heat/core/factories.py

```diff
@@ -117,10 +117,11 @@ def arange(*args, dtype=None, split=None, device=None, comm=MPI_WORLD):
     device = devices.sanitize_device(device)
     data = torch.arange(
         start, stop, step,
-        dtype=types.canonical_heat_type(dtype).torch_type(),
         device=device.torch_device
     )
 
+    data = data.type(types.canonical_heat_type(dtype).torch_type())
```

Comment:
Double inference of heat type (two lines below once again). Since Python does not compile, but is interpreted, this function will actually be executed twice. Introduce a local variable instead and reuse it